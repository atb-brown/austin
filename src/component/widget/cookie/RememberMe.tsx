import React, { useCallback, useEffect } from "react";
import { useCookies } from "react-cookie";
import "../../../common.css";

const clickMeToRemember = "Click me if you want me to remember you!";
const clickMeToForget = "Welcome back! Click me if you want me to forget.";
const clickOk = "; if you're okay with that, click OK.";
const confirmRememberTxt =
  "This will require me to save a cookie on your computer" + clickOk;
const confirmForgetTxt =
  "This will remove the cookie from your computer" + clickOk;

const cookieName = "remembered";
const title =
  "This component allows the site to remember a user. It's an exercise" +
  " in handling cookies and component state!";

/**
 * This button allows the user to be remembered.
 *
 * @return {JSX.Element}
 */
export default function RememberMe(): JSX.Element {
  const [cookies, setCookie, removeCookie] = useCookies([cookieName]);
  const [remember, setRemember] = React.useState(
    Object.hasOwn(cookies, cookieName),
  );
  const [buttonTxt, setButtonTxt] = React.useState(
    remember ? clickMeToRemember : clickMeToForget,
  );
  const [confirmTxt, setConfirmTxt] = React.useState(
    remember ? confirmForgetTxt : confirmRememberTxt,
  );

  const onClick: () => void = useCallback(() => {
    if (confirm(confirmTxt)) {
      if (!remember) {
        setCookie(cookieName, "true");
      } else {
        removeCookie(cookieName);
      }
    }
  }, [confirmTxt, removeCookie, remember, setCookie]);

  useEffect(() => {
    if (Object.hasOwn(cookies, cookieName)) {
      setRemember(true);
      setButtonTxt(clickMeToForget);
      setConfirmTxt(confirmForgetTxt);
    } else {
      setRemember(false);
      setButtonTxt(clickMeToRemember);
      setConfirmTxt(confirmRememberTxt);
    }
  }, [cookies]);
  return (
    <div
      className="clickable"
      data-testid="remember-me"
      onClick={onClick}
      style={{
        alignItems: "center",
        borderRadius: 10,
        display: "flex",
        fontSize: "50%",
        height: 60,
        justifyContent: "center",
        width: "25%",
      }}
      title={title}
    >
      {buttonTxt}
    </div>
  );
}
