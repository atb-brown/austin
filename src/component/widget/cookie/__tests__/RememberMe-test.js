import RememberMe from "../RememberMe";
import "@testing-library/jest-dom";
import { screen, render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Cookies } from "react-cookie";

const cookies = new Cookies();
const confirmSpy = jest.spyOn(window, "confirm");

function confirmOk() {
  confirmSpy.mockImplementation(jest.fn(() => true));
}

function confirmCancel() {
  confirmSpy.mockImplementation(jest.fn(() => false));
}

const testId = "remember-me";
const notRememberedTxt = "Click me if you want me to remember you!";
const rememberedTxt = "Welcome back! Click me if you want me to forget.";
const cookieName = "remembered";

beforeEach(() => {
  confirmOk();
});

it("renders to match snapshot without cookie", () => {
  const tree = renderer.create(<RememberMe />);
  expect(tree).toMatchSnapshot();
});

it("renders to match snapshot with cookie", () => {
  cookies.set(cookieName, true);

  const tree = renderer.create(<RememberMe />);

  expect(tree).toMatchSnapshot();

  cookies.remove(cookieName);
});

it("changes text when clicked and confirmed", () => {
  render(<RememberMe />);

  const rememberMeComp = screen.getByTestId(testId);
  expect(rememberMeComp).toBeDefined();

  // Make sure user is not remembered
  expect(rememberMeComp).toHaveTextContent(notRememberedTxt);

  // Click and confirm and then make sure the user is remembered
  fireEvent.click(rememberMeComp);
  expect(rememberMeComp).toHaveTextContent(rememberedTxt);

  // Click again and make sure the user is not remembered again
  fireEvent.click(rememberMeComp);
  expect(rememberMeComp).toHaveTextContent(notRememberedTxt);
});

it("does not change text when clicked and canceled", () => {
  confirmCancel();
  render(<RememberMe />);

  const rememberMeComp = screen.getByTestId(testId);
  expect(rememberMeComp).toBeDefined();

  // Make sure user is not remembered
  expect(rememberMeComp).toHaveTextContent(notRememberedTxt);

  // Click and cancel and then make sure the user is still not remembered
  fireEvent.click(rememberMeComp);
  expect(rememberMeComp).toHaveTextContent(notRememberedTxt);
});

it("cookie saved when not remembered and button clicked and confirmed", () => {
  render(<RememberMe />);

  const rememberMeComp = screen.getByTestId(testId);
  expect(rememberMeComp).toBeDefined();

  // At first, there shouldn't be a cookie
  expect(cookies.get(cookieName)).toBeUndefined();

  // After clicking and confirming, the cookie should be there.
  fireEvent.click(rememberMeComp);
  expect(cookies.get(cookieName)).toBe(true);

  // And if you click again, the cookie should be removed.
  fireEvent.click(rememberMeComp);
  expect(cookies.get(cookieName)).toBeUndefined();
});

it("cookie is not saved when not remembered and button clicked and canceled", () => {
  confirmCancel();
  render(<RememberMe />);

  const rememberMeComp = screen.getByTestId(testId);
  expect(rememberMeComp).toBeDefined();

  // At first, there shouldn't be a cookie
  expect(cookies.get(cookieName)).toBeUndefined();

  // After clicking and canceling, the cookie should still not be there.
  fireEvent.click(rememberMeComp);
  expect(cookies.get(cookieName)).toBeUndefined();
});
