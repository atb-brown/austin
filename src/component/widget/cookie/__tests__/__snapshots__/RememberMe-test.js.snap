// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders to match snapshot with cookie 1`] = `
<div
  className="clickable"
  data-testid="remember-me"
  onClick={[Function]}
  style={
    {
      "alignItems": "center",
      "borderRadius": 10,
      "display": "flex",
      "fontSize": "50%",
      "height": 60,
      "justifyContent": "center",
      "width": "25%",
    }
  }
  title="This component allows the site to remember a user. It's an exercise in handling cookies and component state!"
>
  Welcome back! Click me if you want me to forget.
</div>
`;

exports[`renders to match snapshot without cookie 1`] = `
<div
  className="clickable"
  data-testid="remember-me"
  onClick={[Function]}
  style={
    {
      "alignItems": "center",
      "borderRadius": 10,
      "display": "flex",
      "fontSize": "50%",
      "height": 60,
      "justifyContent": "center",
      "width": "25%",
    }
  }
  title="This component allows the site to remember a user. It's an exercise in handling cookies and component state!"
>
  Welcome back! Click me if you want me to forget.
</div>
`;
